synapse/federation/federation_base.py

```diff
@@ -80,7 +80,13 @@ async def _check_sigs_and_hash(
           InvalidEventSignatureError if the signature check failed. Nothing
              will be logged in this case.
         """
-        await _check_sigs_on_pdu(self.keyring, room_version, pdu)
+        try:
+            await _check_sigs_on_pdu(self.keyring, room_version, pdu)
+        except Exception as exc:
+            await self.store.record_event_failed_pull_attempt(
+                pdu.room_id, pdu.event_id, str(exc)
```

Comment:
The problem with having this in `_check_sigs_and_hash` is that it runs in situations (like invite, join, knock) where we don't have the room persisted yet and we see `FOREIGN KEY constraint failed`

Reply:
There is potential to move this up to `_check_sigs_and_hash_and_fetch_one` where it's only used for backfill, `get_room_state`, `get_event_auth`, or `get_missing_events` but then we miss the subsequent attempt when we try asking the senders original server via `get_pdu` which goes straight to `_check_sigs_and_hash`.

We also lose the tampered with failure because it doesn't throw an error to the outer function, just returns the redacted event. It's possible that we can see whether it's redacted.

Reply:
Moved to a callback pattern so we can just record in the pulled event scenarios but still get the granularity in types of failures.